### PR TITLE
London | Elhadj Abdoul Diallo| Module-Data-Flows| Week3 - challenges/unit-testing

### PR DESCRIPTION
**Pull request by eediallo**
## Learners, PR Template

Self checklist

- [ ] I have committed my files one by one, on purpose, and for a reason
- [ ] I have titled my PR with COHORT_NAME | FIRST_NAME LAST_NAME | REPO_NAME | WEEK 
- [ ] I have tested my changes
- [ ] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/contributing/)
- [ ] My changes meet the [requirements](./README.md) of this task

## Changelist

Briefly explain your PR.

## Questions

Ask any questions you have for your reviewer.

### CODE REVIEW

File: challenges/unit-testing/katas-tdd/calculator/calculator.test.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+});
+
+
+test("throw and error and display if multiple negative values", () => {
+    const targetOutput = "negatives not allowed: -3, -4";
+    expect(() => {
+      const str = "200000, 1, -3, -4";
+      add(str);
+    }).toThrow(targetOutput);
+  });
```
Comment (mhmdk) [inline challenges/unit-testing/katas-tdd/calculator/calculator.test.js:53]: could you please add tests where the input consists only of a single negative value, or a single large value?

File: challenges/unit-testing/katas-tdd/password-verifier/passwordVerifier.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  const passwordRejected = "Password rejected";
+  if (password == null) {
+    return passwordRejected;
+  }
+
+  if (password.length < 8) {
+    return passwordRejected;
+  }
+
+  if (!includesUpperCaseLetterAndNumber(password)) {
```
Comment (mhmdk) [inline challenges/unit-testing/katas-tdd/password-verifier/passwordVerifier.js:11]: Thanks for the descriptive name!

File: challenges/unit-testing/katas-tdd/roman-numerals/convert-to-old-roman.js [hunk 3]
```diff
@@ -1,3 +1,34 @@
...
+    { value: 1000, numeral: romanNumerals.thousand },
+    { value: 500, numeral: romanNumerals.fiveHundred },
+    { value: 100, numeral: romanNumerals.hundred },
+    { value: 50, numeral: romanNumerals.fifty },
+    { value: 10, numeral: romanNumerals.ten },
+    { value: 5, numeral: romanNumerals.five },
+    { value: 1, numeral: romanNumerals.one },
+  ];
+
+  for (let i = 0; i < numeralValues.length; i++) {
```
Comment (mhmdk) [inline challenges/unit-testing/katas-tdd/roman-numerals/convert-to-old-roman.js:24]: Neat! Notice that this is the same implementation as in the new roman, you just supply it a different numralValues. This is a candidate for a common function

File: challenges/unit-testing/passing-tests/get-average/get-average.js [hunk 4]
```diff
@@ -2,6 +2,13 @@
 // return the average of all the numbers
 // be sure to exclude the strings
 
-function average(numbers) {}
+function average(numbers) {
+  const onlyNumbers = numbers.filter((item) => typeof item === "number");
```
Comment (mhmdk) [inline challenges/unit-testing/passing-tests/get-average/get-average.js:6]: What happens on ["hello"]?

File: challenges/unit-testing/writing-tests/largest-number/largest-number.test.js [hunk 5]
```diff
@@ -1,13 +1,17 @@
...
 
 test("returns largest number in array", function () {
-  // Arrange
-  // Act
-  // Assert
+  const numbers = [3, 21, 88, 4, 36];
+  const currentInput = getLargestNumber(numbers);
+  const targetOutput = 88;
+  expect(currentInput).toEqual(targetOutput);
 });
```
Comment (mhmdk) [inline challenges/unit-testing/writing-tests/largest-number/largest-number.test.js:9]: Please add some more edge cases

File: challenges/unit-testing/writing-tests/remove-vowels/remove-vowels.js [hunk 6]
```diff
@@ -1,29 +1,7 @@
...
-    if (
-      character !== "a" &&
-      character !== "o" &&
-      character !== "i" &&
-      character !== "e" &&
-      character !== "u"
-    ) {
-      result.push(character);
-    }
+function removeVowels(words) {
```
Comment (mhmdk) [inline challenges/unit-testing/writing-tests/remove-vowels/remove-vowels.js:1]: This function is taking an array as well, which makes remove-vowels/remove-vowels-in-array.js pointless. While the code itself is good, could you think of a better structuring?

**Comment by mhmdk** [comment]
@eediallo Amazing work, especially the incremental commit history ;)

File: challenges/unit-testing/katas-tdd/calculator/calculator.test.js (hunk 1 above)
Comment (eediallo) [reply challenges/unit-testing/katas-tdd/calculator/calculator.test.js]: Done!

File: challenges/unit-testing/katas-tdd/password-verifier/passwordVerifier.js (hunk 2 above)
Comment (eediallo) [reply challenges/unit-testing/katas-tdd/password-verifier/passwordVerifier.js:11]: Thanks for the constructive feedback.

File: challenges/unit-testing/katas-tdd/roman-numerals/convert-to-old-roman.js (hunk 3 above)
Comment (eediallo) [reply challenges/unit-testing/katas-tdd/roman-numerals/convert-to-old-roman.js:24]: I totally agree. I just did it because they have two files in the folder(the old and the new) but the new roman implementation is enough.  Do you think I should just delete the old one?

File: challenges/unit-testing/passing-tests/get-average/get-average.js (hunk 4 above)
Comment (eediallo) [reply challenges/unit-testing/passing-tests/get-average/get-average.js:6]: ```hello``` is ignored because it is not a number.

File: challenges/unit-testing/writing-tests/largest-number/largest-number.test.js (hunk 5 above)
Comment (eediallo) [reply challenges/unit-testing/writing-tests/largest-number/largest-number.test.js:9]: Done!

File: challenges/unit-testing/writing-tests/remove-vowels/remove-vowels.js (hunk 6 above)
Comment (eediallo) [reply challenges/unit-testing/writing-tests/remove-vowels/remove-vowels.js:1]: Sure!

**Comment by eediallo** [comment]
@mhmdk thank you for your constructive feedback! I really appreciate it.

File: challenges/unit-testing/katas-tdd/roman-numerals/convert-to-old-roman.js (hunk 3 above)
Comment (mhmdk) [reply challenges/unit-testing/katas-tdd/roman-numerals/convert-to-old-roman.js:24]: Not delete, in these cases you create a third file, put the function convertToOldRoman there, and the original two files can just initialize the arguments, and call it.
This is a stretch and not required for this module. Your implementation is already very good

File: challenges/unit-testing/passing-tests/get-average/get-average.js (hunk 4 above)
Comment (mhmdk) [reply challenges/unit-testing/passing-tests/get-average/get-average.js:6]: Avoid dividing by zero. In some other programming languages it would lead to a crash

File: challenges/unit-testing/katas-tdd/calculator/calculator.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  let sum = 0;
+  if (typeof numbers !== "string") {
+    throw new Error("Parameter must be a string.");
+  }
+
+  if (numbers === "") {
+    return sum;
+  }
+
+  if (numbers.length === 1) {
```
Comment (mhmdk) [inline challenges/unit-testing/katas-tdd/calculator/calculator.js:11]: Do yo still need this special case?

File: challenges/unit-testing/writing-tests/remove-vowels/remove-vowels-in-array.js [hunk 8]
```diff
@@ -1,7 +1,7 @@
-let removeVowels = require("./remove-vowels");
-
 function removeVowelsFromWords(words) {
-  return removeVowels(words);
+  return words.map(function (word) {
+    return word.replace(/[aeiou]/gi, "");
```
Comment (mhmdk) [inline challenges/unit-testing/writing-tests/remove-vowels/remove-vowels-in-array.js:3]: Ok this also works. FYI what I would have done: keep this line "word.replace(..)" in remove-vowels.js, so that it removes vowels from a single word, no array. And here I can call words.map(removeVowels).
The point is having the function in this file take an array, and returning an array. And having the function in remove-vowels.js take a string, and return a string.
But again you code is good enough

File: challenges/unit-testing/katas-tdd/calculator/calculator.js (hunk 7 above)
Comment (eediallo) [reply challenges/unit-testing/katas-tdd/calculator/calculator.js:11]: Yes, I think I still need it because of the way I have written the code. Is there a better way to handle this?

Comment (mhmdk) [reply challenges/unit-testing/katas-tdd/calculator/calculator.js:11]: I thought both this, and the check on line 16 can be removed together. May be try and check if any test will fail.

Comment (eediallo) [reply challenges/unit-testing/katas-tdd/calculator/calculator.js:11]: remove this check will make this test fail ```test("return number if param length is 1", () => {
  const str = "9";
  const currentInput = add(str);
  const targetOutput = 9;
  expect(currentInput).toEqual(targetOutput);
});```

Comment (eediallo) [reply challenges/unit-testing/katas-tdd/calculator/calculator.js]: I have removed both checks and all test cases passed. Thank you!

Comment (mhmdk) [reply challenges/unit-testing/katas-tdd/calculator/calculator.js]: yep, with TDD usually you will need these kind of conditions to make the tests pass, but once green you can look for simplifications

Comment (eediallo) [reply challenges/unit-testing/katas-tdd/calculator/calculator.js]: I see! Thank you very much!
